File: SHOW_AND_TELL/SERIALIZATION_6_12_2019/examples/ex.swift
struct Parent: Codable {
  var id:Int
  var foo:String
  var baz:String

  var children:[Child]?
}

struct Child: Codable {
  var id:Int
  var key:String
  var optionalString:String?
}

class ApiClient {
  func getParentWithId(_ id: Int, completion: (err: Error, result: Parent?) -> Void) {
    let json = get("/Parents/\(id)") as? data else { completion(Error(), nil) }
    let decoder = JSONDecoder()
    if let parent = try decoder.decode(Parent.self, from: json) {
      completion(parent)
    } else {
      completion(error)
    }
  }
}

func getParents() {
  for id in self.parentIds {
    ApiClient.getParentWithId(id, { err, parent
      // check err or do thing with parent
    })
  }
}

File: SHOW_AND_TELL/SERIALIZATION_6_12_2019/examples/ex.js
// In order to create an instance of an object in js, 
// we use a variety of clients and custom solutions.
// i.e. redux, orms, etc

class Parent {
  constructor(data) {
    this.id = data.id
    this.foo = data.foo
    this.baz = data.baz
    this._children = data.children
  }

  get children() {
    this.children.map(child => { new Child(child) })
  }

  set children(children) {
    this._children = children
  }
}

class Child {
  constructor(data) {
    this.id = data.id
    this.key = data.key
    this.optionalString = data.optionalString
  }

  get optionalString() {
    this.optionalString ? this.optionalString : undefined
  }
}
